src/main/java/com/direwolf20/buildinggadgets/client/RemoteInventoryCache.java

```diff
@@ -0,0 +1,76 @@
+package com.direwolf20.buildinggadgets.client;
+
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.Nullable;
+
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.Pair;
+
+import com.direwolf20.buildinggadgets.common.network.PacketHandler;
+import com.direwolf20.buildinggadgets.common.network.PacketSetRemoteInventoryCache;
+import com.direwolf20.buildinggadgets.common.tools.GadgetUtils;
+import com.direwolf20.buildinggadgets.common.tools.UniqueItem;
+import com.google.common.base.Stopwatch;
+import com.google.common.collect.Multiset;
+
+import net.minecraft.item.ItemStack;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.util.math.BlockPos;
+import net.minecraftforge.fml.relauncher.Side;
+import net.minecraftforge.fml.relauncher.SideOnly;
+
+@SideOnly(Side.CLIENT)
+public class RemoteInventoryCache {
+    private boolean isCopyPaste;
+    private Pair<Integer, BlockPos> locCached;
+    private Multiset<UniqueItem> cache;
+    private Stopwatch timer;
```

Comment:
Why not use a Guava Cache?

Reply:
The thing with Guava caches is that you need to be able to statically compute the values from only the keys, and this is done automatically when attempting to get a cached value that doesn't exist. In this case, the values are computed in a packet on the server from both the keys and the packet's context, then sent to the client.

Just to see, I implemented it with a Guava cache anyway, and I found that it takes about the same amount of code as the current Multiset/Stopwatch implementation, but without the nice Multiset functionality (meaning more code in InventoryManipulation). Ultimately, I think the current way is better.